cfgrammar/src/lib/yacc/ast.rs

```diff
@@ -301,6 +304,19 @@ impl GrammarAST {
         Ok(())
     }
 
+    pub fn warnings(&self) -> impl Iterator<Item = YaccGrammarWarning> + '_ {
```

Comment:
I wonder if having this be an iterator is useful in the sense that callers will always `collect` it? [I don't have a strong opinion here -- this is a genuine question!]

Reply:
Yeah, I don't know if it helps here, I think that just ends up having to `collect` it twice, in the case where `YaccGrammar::new_with_validation`  takes ownership, and these generally need to outlive that until the `chain` call.

I also find the iterator here sketchy too though, and much preferred the way that the `Analysis` could take ownership while avoiding making any iterator public.

Reply:
>I also find the iterator here sketchy too though, and much preferred the way that the Analysis could take ownership while avoiding making any iterator public.

Could you expand this thought?

Reply:
I find it sketchy mostly because of that lifetime, what if another type of warning comes along and it is a pain to chain it to the `unused_symbol_warnings().map(...)` due to some lifetime requirement. I would much prefer as you indicated above to `collect` and have an owned `Vec<YaccGrammarWarning>` somewhere.  The complication is the lifetimes of it all.

With `YaccGrammar` taking ownership of `ASTWithValidityInfo` and essentially dropping it before return,
when we want `warnings()` beyond that (when we want to report both warnings and errors).

 `Analysis` was nice in that it was owned by the caller, and consumed `warnings()` via a `&mut self`. So `Analysis` could outlive `ASTWithValidityInfo` after the latter was consumed by a `YaccGrammar`.

Reply:
`GrammarAST` could have a `warnings: Vec<...>` field or similar perhaps? Or we could make the `warning`s function `crate` only and external users have to find this info out via `ASTWithValidityInfo`?

Reply:
It'd have to be `YaccGrammar` having the warnings field, because `GrammarAST` doesn't live long enough either.
But that doesn't actually work in the case where `YaccGrammarResult` returns an `Err`. 

One thing we could perhaps do though is make it `fn get_warnings(&mut Vec<YaccGrammarWarning>)`?

Reply:
Maybe there's something implicit here: I think you want the warnings identified at AST-time to live after the AST has been consumed and a `YaccGrammar` created?

Reply:
Yes, exactly the AST only lives until `new_with_validation()`, while all the `warnings.is_empty()` and iteration, are done
matching on the result of that call.

Reply:
I'm not sure what I was thinking here, because we don't store `warnings` as a field in `ASTWithValidityInfo`,
But it is a value derived from `GrammarAST`, we don't end up having to clone the vector, to extend it's lifetime, e.g.
such if we were trying to keep `errors` field around after the ast has been moved.

So it seems fine to just collect in the `warnings()` function and return/move it?

I went ahead and commited that in [8530c8c](https://github.com/softdevteam/grmtools/pull/354/commits/8530c8c05371999fcc155e1f108bdee9cbf1d1b2)